File: 13302/Main.cpp
/*
    ref by: https://www.acmicpc.net/board/view/25480
*/

#include<iostream>
#include<algorithm>
using namespace std;

void print2DArray(int t[][50], int N);

int main()
{
	int t[106][50], M[101];
	int i, j, o, m, N;
	cin >> N >> m;
	while (m--){
		cin >> M[0];
		M[M[0]] = -1;
	}
	for (i = 0; i <= N; i++)
		for (j = 0; j <= 50; j++)
			t[i][j] = 1e9;
	t[0][0] = 0;
	for (i = 0; i <= N; i++) {
		for (j = 0; j <= 43; j++)
		{
			if (M[i + 1] == -1)t[i + 1][j] = min(t[i][j], t[i + 1][j]);

			t[i + 1][j] = min(t[i + 1][j], t[i][j] + 10000);
			t[i + 3][j + 1] = min(t[i + 3][j + 1], t[i][j] + 25000);
			t[i + 5][j + 2] = min(t[i + 5][j + 2], t[i][j] + 37000);

			for (o = 3; o <= j; o++)
				t[i + 1][j - 3] = min(t[i + 1][j - 3], t[i][j]);
		}
	}
	m = t[N][0];
	for (i = 1; i < 43; i++)
		if (m > t[N][i])
			m = t[N][i];
	cout << m;
    // print2DArray(t, N);

	return 0;
}

void print2DArray(int t[][50], int N){
    for(int i=0;i<N;i++){
        for(int j=0;j<43;j++){
            if(t[i][j] != 1e9)
                cout<<t[i][j]<<" ";
        }
        cout<<endl;
    }
}

File: 13302/Main.java
/*
    baekjoon online judge
    problem number 13302
    https://www.acmicpc.net/problem/13302
    using dynaimc programming
    https://www.acmicpc.net/board/view/25480
*/
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.OutputStreamWriter;
import java.io.BufferedWriter;
import java.util.StringTokenizer;
import java.io.IOException;
import java.util.Arrays;

public class Main{
    public static void main(String[] args)throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[] abDays = new int[M];
        // this line is very important when M is 0
        if(M>0){
            st = new StringTokenizer(br.readLine());
            for(int i=0;i<M;i++){
                abDays[i] = Integer.parseInt(st.nextToken());
            }
        }
        int price = calculateDays(N, abDays);

        bw.write(Integer.toString(price)+"\n");
        bw.flush();
        bw.close();
        br.close();
    }

    // plan: row for days, column for coupon numbers
    public static int calculateDays(int N, int[] abDays){
        int[][] plan = new int[110][50];
        for(int i=0;i<plan.length;i++)
            Arrays.fill(plan[i], 10000000);
        plan[0][0] = 0;

        for(int i=0;i<N;i++){
            for(int j=0;j<48;j++){
                // if day is absent days than skip it
                if(isAbsentDays(i+1, abDays))
                    plan[i+1][j] = MIN(plan[i+1][j],plan[i][j]);
                // calculate 3 cases
                plan[i+1][j] = MIN(plan[i+1][j], plan[i][j]+10000);
                plan[i+3][j+1] = MIN(plan[i+3][j+1], plan[i][j]+25000);
                plan[i+5][j+2] = MIN(plan[i+5][j+2], plan[i][j]+37000);
            
                // if have enough coupon also calculate it
                if(j >= 3)
                    plan[i+1][j-3] = MIN(plan[i+1][j-3], plan[i][j]);
            }
        }

        int price = minPrice(plan[N]);
        return price;
    }

    public static boolean isAbsentDays(int day, int[] days){
        for(int i=0;i<days.length;i++){
            if(days[i] == day)
                return true;
            else if(days[i] > day)
                break;
        }
        return false;
    }

    public static int minPrice(int[] arr){
        int min = 10000000;

        for(int v: arr){
            if(min>v)
                min = v;
        }
        return min;
    }

    public static int MIN(int a, int b){
        return a<b?a:b;
    }
}
